tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc

```diff
@@ -0,0 +1,132 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA && GOOGLE_TENSORRT
+
+#include "tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/op_converter_registry.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h"
+
+namespace tensorflow {
+namespace tensorrt {
+namespace convert {
+
+class ConvertSelectV2 : public OpConverterBase<ConvertSelectV2> {
+ public:
+  explicit ConvertSelectV2(OpConverterParams* params)
+      : OpConverterBase<ConvertSelectV2>(params) {}
+
+  static constexpr std::array<DataType, 3> AllowedDataTypes() {
+    return {DataType::DT_FLOAT, DataType::DT_HALF, DataType::DT_INT32};
+  }
+
+  static constexpr std::array<InputArgSpec, 3> InputSpec() {
+    return std::array<InputArgSpec, 3>{
+        InputArgSpec::Create("cond", TrtInputArg::kBoth),
+        InputArgSpec::Create("then", TrtInputArg::kBoth),
+        InputArgSpec::Create("else", TrtInputArg::kBoth)};
+  }
+
+  Status Validate() {
+    const auto& params = *this->params_;
+    if (params.use_implicit_batch) {
+      return errors::Unimplemented(
+          "Conversion for SelectV2 is not supported in implicit batch mode.");
+    }
+
+    const auto& inputs = params.inputs;
+    const TRT_TensorOrWeights& cond_input = inputs.at(0);
+    const auto& node = params.node_def;
+    TF_RETURN_IF_ERROR(
+        check_type(cond_input.TrtDType(), nvinfer1::DataType::kBOOL, node));
+
+    const auto type_then = inputs[1].TrtDType();
+    const auto type_else = inputs[2].TrtDType();
+    if (type_then != type_else && (type_then == nvinfer1::DataType::kINT32 ||
+                                   type_else == nvinfer1::DataType::kINT32)) {
+      // Both or none of (type_then, type_else) should be equal to kINT32.
+      return errors::InvalidArgument(
+          then_else_dtypes_error_msg(type_then, type_else, node));
+    }
+
+    all_tensors_ = true;
+    for (int i = 0; i < 3; i++) {
+      if (inputs.at(i).is_weights()) {
+        if (!i) {
+          return errors::InvalidArgument(bool_weight_error_msg(node));
+        }
+        all_tensors_ = false;
+        break;
+      }
+    }
+
+    nvinfer1::Dims broadcasted_dims[3];
+    for (int i = 1; i < 3; i++) {
+      TF_RETURN_IF_ERROR(GetTrtBroadcastShape(inputs.at(0), inputs.at(i), true,
+                                              false, broadcasted_dims,
+                                              broadcasted_dims + i));
```

Comment:
Will this work with the following input shapes [2, 1, 1], [1, 2, 1], [1, 1, 2]?

- I believe you will need to reconcile broadcasted_dims[1] and proadcasted_dims[2], and then adjust broadcasted_dims[0] accordingly.
- Alternatively you could add on optional third input dim for GetTrtBroadcastShape.

Reply:
Unfortunately, it is impossible to reconcile `broadcasted_dims[1]` and `broadcasted_dims[2]`, when both `then` and `else` parameters are passed as weights. On the other hand, today we allow passing `cond` only as a tensor, I am using 
```
GetTrtBroadcastShape(inputs.at(0), inputs.at(i), ...
```

Reply:
Then we should make it possible to reconcile when both `then` and `else` are weights.

Reply:
Yes, it is possible and we have [a subtest](https://github.com/tensorflow/tensorflow/blob/690d347af047876756c492024dd502f7497826ee/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L9316-L9324) for this, which checks all possible combinations of the passed parameters.